### PR TITLE
Add fungible token class

### PR DESCRIPTION
**Pull request by DevRozaDev**

### CODE REVIEW

File: src/fungible_token.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+export class FungibleToken {
+  network: Network
+  nodeUrl: string
+  builder: TransactionBuilder
+
+  constructor(network: Network) {
+    this.network = network
+
+    if (network === Network.Local || network === Network.Devnet) {
```
Comment (Sniezka1927) [inline src/fungible_token.ts:15]: I think it is worth to extract this logic to separate function

File: src/fungible_token.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    const buildTxResult = await this.mintTx(signer, value, tokenAddress)
+    const signerAddress = (await signer.getSelectedAccount()).address
+    return await signer.signAndSubmitUnsignedTx({
+      signerAddress: signerAddress,
+      unsignedTx: buildTxResult.unsignedTx
+    })
+  }
+
+  async getAllBalances(tokens: string[], owner: string): Promise<Map<string, bigint>> {
+    const balancePromises = await Promise.all(tokens.map(token => this.balanceOf(owner, token)))
```
Comment (Sniezka1927) [inline src/fungible_token.ts:85]: I think the promises are already resolved right
```suggestion
    const balances = await Promise.all(tokens.map(token => this.balanceOf(owner, token)))
```

**Review by Sniezka1927 — CHANGES_REQUESTED** [review]

File: src/fungible_token.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    symbol: string = '',
+    decimals: bigint = 0n
+  ): Promise<string> {
+    const deployResult = await waitTxConfirmed(
+      TokenFaucet.deploy(signer, {
+        initialFields: {
+          name: Buffer.from(name, 'utf8').toString('hex'),
+          symbol: Buffer.from(symbol, 'utf8').toString('hex'),
+          decimals,
+          supply,
```
Comment (Sniezka1927) [inline src/fungible_token.ts:31]: If we are implementing this design the totalSupply should be also set to max u256

**Review by Sniezka1927 — CHANGES_REQUESTED** [review]
In contract we are using `contractId` more often than token address. There should be also a getter for that

**Review by Sniezka1927 — APPROVED** [review]

File: src/fungible_token.ts [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+    return new Map(tokens.map((token, i) => [token, balances[i]]))
+  }
+
+  async balanceOf(owner: string, tokenAddress: string): Promise<bigint> {
+    const tokenId = this.getContractId(tokenAddress)
+    return balanceOf(tokenId, owner)
+  }
+
+  async tokenName(tokenAddress: string): Promise<string> {
```
Comment (wojciech-cichocki) [inline src/fungible_token.ts:89]: Is it possible to query all token metadata in one query without using `Promse.all` but through view itself?

Comment (DevRozaDev) [reply src/fungible_token.ts:89]: Yes, using `multicall`, will be added.

**Comment by DevRozaDev** [comment]
```ts
await token.multicall({ getSymbol: {}, getName: {}, getDecimals: {} })
```
gives different result than
```ts
await token.multicall({ getName: {}, getSymbol: {}, getDecimals: {} })
```
despite of us accessing the results using a `.methodName.results` syntax.
Till this is addressed upstream we should be vary of `multicall` and thoroughly check what it is returning.

**Review by Sniezka1927 — APPROVED** [review]

File: src/fungible_token.ts [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+      decimals: result.getDecimals.returns
+    }
+  }
+
+  async tokenName(tokenAddress: string): Promise<string> {
+    const token = TokenFaucet.at(tokenAddress)
+    return hexToString((await token.view.getName()).returns)
+  }
+
+  async tokenSymbol(tokenAddress: string): Promise<string> {
```
Comment (wojciech-cichocki) [inline src/fungible_token.ts:118]: should be named `getTokenSymbol`. Same should be in other places in `FungibleToken` class

File: src/fungible_token.ts [hunk 6]
```diff
@@ -0,0 +1,131 @@
+import { DUST_AMOUNT, hexToString, SignerProvider, TransactionBuilder } from '@alephium/web3'
```
Comment (wojciech-cichocki) [inline src/fungible_token.ts:1]: Check the naming convention for files in TS. Should be `fungible-token.ts`. Same case in `test/fungible_token.test.ts`

**Review by wojciech-cichocki — APPROVED** [review]

**Pull request merged**